File: Sources/APTradeApp/PlansViewModel.swift
import Foundation
import APTradeApplication
import APTradeDomain

/// Plans tab: lists the user's investment Pies as summary cards and drives the detail
/// sheet, one-off contributions, and manual rebalancing. Every user-visible string this
/// view model produces (the insufficient-cash error, the next-contribution label) is
/// routed through `tr(_:)`.
@MainActor
@Observable
public final class PlansViewModel {

    /// One list-card's worth of display data for a Pie.
    public struct PieRow: Identifiable, Equatable {
        public let id: String
        public let name: String
        public let currentValue: Money
        public let nextContributionLabel: String?
        /// Largest absolute drift (percentage points) across the pie's slices. The view
        /// shows a drift badge when this exceeds 5pp.
        public let maxDriftPP: Decimal
        public let sliceWeights: [(String, Percentage)]

        public init(id: String, name: String, currentValue: Money, nextContributionLabel: String?,
                   maxDriftPP: Decimal, sliceWeights: [(String, Percentage)]) {
            self.id = id
            self.name = name
            self.currentValue = currentValue
            self.nextContributionLabel = nextContributionLabel
            self.maxDriftPP = maxDriftPP
            self.sliceWeights = sliceWeights
        }

        // `[(String, Percentage)]` can't derive `Equatable` (tuples don't conform to the
        // protocol), so this compares element-by-element instead.
        public static func == (lhs: PieRow, rhs: PieRow) -> Bool {
            lhs.id == rhs.id
                && lhs.name == rhs.name
                && lhs.currentValue == rhs.currentValue
                && lhs.nextContributionLabel == rhs.nextContributionLabel
                && lhs.maxDriftPP == rhs.maxDriftPP
                && lhs.sliceWeights.elementsEqual(rhs.sliceWeights) { $0.0 == $1.0 && $0.1 == $1.1 }
        }
    }

    /// Full detail-sheet data for one Pie: each slice's target vs. actual allocation and
    /// drift, the activity log, and the recurring contribution schedule (if any).
    public struct PieDetail: Equatable {
        public struct SliceDetail: Identifiable, Equatable {
            public let symbol: String
            public let assetKind: AssetKind
            public let targetWeight: Percentage
            public let actualWeight: Percentage
            public let drift: Percentage
            public let currentValue: Money
            public var id: String { symbol }

            public init(symbol: String, assetKind: AssetKind, targetWeight: Percentage,
                       actualWeight: Percentage, drift: Percentage, currentValue: Money) {
                self.symbol = symbol
                self.assetKind = assetKind
                self.targetWeight = targetWeight
                self.actualWeight = actualWeight
                self.drift = drift
                self.currentValue = currentValue
            }
        }

        public let pieId: String
        public let name: String
        public let slices: [SliceDetail]
        public let activity: [PieActivityEntry]
        public let schedule: ContributionSchedule?

        public init(pieId: String, name: String, slices: [SliceDetail],
                   activity: [PieActivityEntry], schedule: ContributionSchedule?) {
            self.pieId = pieId
            self.name = name
            self.slices = slices
            self.activity = activity
            self.schedule = schedule
        }
    }

    public private(set) var rows: [PieRow] = []
    public private(set) var detail: PieDetail?
    public private(set) var rebalancePreview: [RebalanceOrder]?
    public private(set) var errorMessage: String?

    private let loadPies: LoadPies
    private let deletePieUseCase: DeletePie
    private let contributeToPie: ContributeToPie
    private let rebalancePie: RebalancePie
    private let reconcileLedgers: ReconcilePieLedgers
    private let fetchQuotes: FetchQuotesUseCase
    private let calendar: MarketCalendar
    private let now: () -> Date

    public init(
        loadPies: LoadPies,
        deletePie: DeletePie,
        contributeToPie: ContributeToPie,
        rebalancePie: RebalancePie,
        reconcileLedgers: ReconcilePieLedgers,
        fetchQuotes: FetchQuotesUseCase,
        calendar: MarketCalendar = MarketCalendar(),
        now: @escaping () -> Date = Date.init
    ) {
        self.loadPies = loadPies
        self.deletePieUseCase = deletePie
        self.contributeToPie = contributeToPie
        self.rebalancePie = rebalancePie
        self.reconcileLedgers = reconcileLedgers
        self.fetchQuotes = fetchQuotes
        self.calendar = calendar
        self.now = now
    }

    /// Reconciles every Pie's ledger against actual portfolio holdings BEFORE building
    /// rows, so a manual sell made outside a Pie can never show stale, over-claimed
    /// values here (see `ReconcilePieLedgers`'s cross-pie drift note).
    public func onAppear() async {
        _ = reconcileLedgers()
        await reloadRows()
    }

    public func openDetail(id: String) async {
        rebalancePreview = nil // a preview from a different pie must never linger
        guard let pie = loadPies().first(where: { $0.id == id }) else {
            detail = nil
            return
        }
        detail = await buildDetail(for: pie)
    }

    public func contributeNow(id: String, amount: Money) async {
        errorMessage = nil
        rebalancePreview = nil
        guard amount.amount > 0 else {
            errorMessage = tr(.pieInvalidAmount)
            return
        }
        let day = calendar.tradingDay(of: now())
        do {
            let outcome = try await contributeToPie(pieId: id, amount: amount, day: day, now: now())
            if case .skippedInsufficientCash = outcome {
                errorMessage = tr(.pieInsufficientCash)
            }
            await reloadRows()
            if detail?.pieId == id { await openDetail(id: id) }
        } catch is CancellationError {
            // Cooperative cancellation: leave state untouched rather than surfacing it
            // as a contribution failure (mirrors CalendarViewModel's convention).
        } catch {
            errorMessage = tr(.couldntPlaceOrder)
        }
    }

    /// Prices the Pie's current holdings against live quotes and fills `rebalancePreview`
    /// with the trades `confirmRebalance` would place. Read-only.
    public func requestRebalance(id: String) async {
        errorMessage = nil
        do {
            rebalancePreview = try await rebalancePie.preview(pieId: id)
        } catch is CancellationError {
        } catch {
            errorMessage = tr(.couldntPlaceOrder)
        }
    }

    /// Executes the previewed orders, then clears the preview and refreshes rows/detail.
    public func confirmRebalance(id: String) async {
        errorMessage = nil
        do {
            _ = try await rebalancePie.execute(pieId: id, day: calendar.tradingDay(of: now()), now: now())
            rebalancePreview = nil
            await reloadRows()
            if detail?.pieId == id { await openDetail(id: id) }
        } catch is CancellationError {
        } catch {
            errorMessage = tr(.couldntPlaceOrder)
        }
    }

    public func deletePie(id: String) async {
        rebalancePreview = nil
        _ = deletePieUseCase(id: id)
        if detail?.pieId == id { detail = nil }
        await reloadRows()
    }

    // MARK: - Row / detail construction

    private func reloadRows() async {
        let pies = loadPies()
        let symbols = Array(Set(pies.flatMap { $0.slices.map(\.symbol) }))
        let quotes = await fetchQuotes(symbols: symbols)
        rows = pies.map { buildRow(for: $0, quotes: quotes) }
    }

    private func buildRow(for pie: Pie, quotes: [String: Result<Quote, AppError>]) -> PieRow {
        let currentValues = currentValues(for: pie, quotes: quotes)
        let totalValue = currentValues.values.reduce(Money(amount: 0)) { $0 + $1 }
        let drift = PieMath.drift(currentValues: currentValues, targets: pie.slices)
        let maxDriftPP = drift.values.map { abs($0.value) }.max() ?? 0
        return PieRow(
            id: pie.id,
            name: pie.name,
            currentValue: totalValue,
            nextContributionLabel: nextContributionLabel(for: pie),
            maxDriftPP: maxDriftPP,
            sliceWeights: pie.slices.map { ($0.symbol, $0.targetWeight) }
        )
    }

    private func buildDetail(for pie: Pie) async -> PieDetail {
        let quotes = await fetchQuotes(symbols: pie.slices.map(\.symbol))
        let currentValues = currentValues(for: pie, quotes: quotes)
        let totalValue = currentValues.values.reduce(Decimal(0)) { $0 + $1.amount }
        let driftBySymbol = PieMath.drift(currentValues: currentValues, targets: pie.slices)
        let sliceDetails = pie.slices.map { slice -> PieDetail.SliceDetail in
            let value = currentValues[slice.symbol] ?? Money(amount: 0)
            let actual = totalValue > 0
                ? Percentage(value: (value.amount / totalValue) * 100)
                : Percentage(value: 0)
            return PieDetail.SliceDetail(
                symbol: slice.symbol,
                assetKind: slice.assetKind,
                targetWeight: slice.targetWeight,
                actualWeight: actual,
                drift: driftBySymbol[slice.symbol] ?? Percentage(value: 0),
                currentValue: value
            )
        }
        return PieDetail(pieId: pie.id, name: pie.name, slices: sliceDetails, activity: pie.activity, schedule: pie.schedule)
    }

    /// Prices `pie`'s ledger against `quotes`; a symbol with no successful quote is
    /// simply omitted (mirrors `PieMath`'s own "0 if missing" treatment).
    private func currentValues(for pie: Pie, quotes: [String: Result<Quote, AppError>]) -> [String: Money] {
        var result: [String: Money] = [:]
        for slice in pie.slices {
            guard case .success(let quote)? = quotes[slice.symbol] else { continue }
            let quantity = pie.quantity(of: slice.symbol)
            result[slice.symbol] = Money(amount: quantity.amount * quote.price.amount, currencyCode: quote.price.currencyCode)
        }
        return result
    }

    private func nextContributionLabel(for pie: Pie) -> String? {
        guard let schedule = pie.schedule,
              let date = PieSchedule.date(fromDay: schedule.nextDueDay, calendar: calendar)
        else { return nil }
        return String(format: tr(.nextContributionFormat), Self.dueDayFormatter.string(from: date))
    }

    private static let dueDayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US")
        f.dateFormat = "MMM d"
        return f
    }()
}

File: desktopApp/src/main/kotlin/com/aptrade/desktop/plans/PlansViewModel.kt
package com.aptrade.desktop.plans

import com.aptrade.desktop.l10n.tr
import com.aptrade.desktop.l10n.trf
import com.aptrade.shared.application.ContributeToPie
import com.aptrade.shared.application.ContributionOutcome
import com.aptrade.shared.application.DeletePie
import com.aptrade.shared.application.FetchMarketQuotes
import com.aptrade.shared.application.LoadPies
import com.aptrade.shared.application.QuoteError
import com.aptrade.shared.application.ReconcilePieLedgers
import com.aptrade.shared.application.RebalancePie
import com.aptrade.shared.domain.AssetKind
import com.aptrade.shared.domain.ContributionSchedule
import com.aptrade.shared.domain.MONEY_MATH
import com.aptrade.shared.domain.MarketCalendar
import com.aptrade.shared.domain.Money
import com.aptrade.shared.domain.Pie
import com.aptrade.shared.domain.PieActivityEntry
import com.aptrade.shared.domain.PieMath
import com.aptrade.shared.domain.PieSchedule
import com.aptrade.shared.domain.Quote
import com.aptrade.shared.domain.RebalanceOrder
import com.aptrade.shared.l10n.L10n
import com.ionspin.kotlin.bignum.decimal.BigDecimal
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

private val ZERO: BigDecimal = BigDecimal.ZERO

/** One list-card's worth of display data for a Pie. `currentValue` and `sliceWeights`
 *  (target-weight percentage points) stay RAW domain types — same "feeds further
 *  formatting" contract [PortfolioViewModel]'s `marketValueText` establishes — the view
 *  (Task 12) is responsible for money/percent text rendering. `nextContributionLabel` is
 *  the one field this VM pre-localizes, since it's a full sentence (`tr`/`trf`), not a
 *  formattable number — same split [PortfolioViewModel] draws between raw amounts and
 *  pre-formatted signed/labeled text. */
data class PieRowUi(
    val id: String,
    val name: String,
    val currentValue: Money,
    val nextContributionLabel: String?,
    /** Largest absolute drift (percentage points) across the pie's slices. The view shows
     *  a drift badge when this exceeds 5pp. */
    val maxDriftPP: BigDecimal,
    val sliceWeights: List<Pair<String, BigDecimal>>,
)

/** One slice's target-vs-actual allocation and drift, for the detail screen. */
data class PieSliceDetailUi(
    val symbol: String,
    val assetKind: AssetKind,
    val targetWeight: BigDecimal,
    val actualWeight: BigDecimal,
    val drift: BigDecimal,
    val currentValue: Money,
)

/** Full detail-screen data for one Pie. `totalValue` is the SAME reduction [PieRowUi]'s
 *  `currentValue` derives for this pie's list card — one source of truth for "how much is
 *  this pie worth" shared by the list and detail, never re-derived by the view. */
data class PieDetailUi(
    val pieId: String,
    val name: String,
    val slices: List<PieSliceDetailUi>,
    val activity: List<PieActivityEntry>,
    val schedule: ContributionSchedule?,
    val totalValue: Money,
)

data class PlansUiState(
    val rows: List<PieRowUi> = emptyList(),
    val detail: PieDetailUi? = null,
    val rebalancePreview: List<RebalanceOrder>? = null,
    val errorMessage: String? = null,
)

/** Plans tab: lists the user's investment Pies as summary cards and drives the detail
 *  screen, one-off contributions, and manual rebalancing. Transcribed from Swift
 *  `PlansViewModel` (`Sources/APTradeApp/PlansViewModel.swift`) AS-BUILT.
 *
 *  Every user-visible string this VM produces (the insufficient-cash error, the
 *  next-contribution label) is routed through `tr`/`trf` (`com.aptrade.desktop.l10n`),
 *  mirroring Swift's `tr(_:)` calls exactly.
 *
 *  Follows the house desktop-VM convention (see [PortfolioViewModel]/[CalendarViewModel]/
 *  [DetailViewModel]): a single [MutableStateFlow]-backed [state], public methods are
 *  plain (non-suspend) event handlers that internally `scope.launch`, and `scope` MUST be
 *  single-thread-confined (Dispatchers.Main on desktop) — the same contract every other
 *  desktop VM in this codebase relies on instead of locks. */
class PlansViewModel(
    private val loadPies: LoadPies,
    private val deletePieUseCase: DeletePie,
    private val contributeToPie: ContributeToPie,
    private val rebalancePie: RebalancePie,
    private val reconcileLedgers: ReconcilePieLedgers,
    private val fetchMarketQuotes: FetchMarketQuotes,
    private val calendar: MarketCalendar,
    private val scope: CoroutineScope,
    private val nowEpochSeconds: () -> Long,
) {
    private val _state = MutableStateFlow(PlansUiState())
    val state: StateFlow<PlansUiState> = _state

    /** Reconciles every Pie's ledger against actual portfolio holdings BEFORE building
     *  rows, so a manual sell made outside a Pie can never show stale, over-claimed
     *  values here (see [ReconcilePieLedgers]'s cross-pie drift note). */
    fun onAppear() {
        scope.launch {
            reconcileLedgers.execute(nowEpochSeconds())
            reloadRows()
        }
    }

    fun openDetail(id: String) {
        // A preview from a different pie must never linger.
        _state.update { it.copy(rebalancePreview = null) }
        scope.launch { loadDetail(id) }
    }

    fun contributeNow(id: String, amount: Money) {
        _state.update { it.copy(errorMessage = null, rebalancePreview = null) }
        if (amount.amount <= ZERO) {
            _state.update { it.copy(errorMessage = tr(L10n.Key.PieInvalidAmount)) }
            return
        }
        scope.launch {
            val day = calendar.tradingDay(nowEpochSeconds())
            try {
                val outcome = contributeToPie.execute(id, amount, day, nowEpochSeconds())
                if (outcome is ContributionOutcome.SkippedInsufficientCash) {
                    _state.update { it.copy(errorMessage = tr(L10n.Key.PieInsufficientCash)) }
                }
                reloadRows()
                if (_state.value.detail?.pieId == id) loadDetail(id)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                _state.update { it.copy(errorMessage = tr(L10n.Key.CouldntPlaceOrder)) }
            }
        }
    }

    /** Prices the Pie's current holdings against live quotes and fills `rebalancePreview`
     *  with the trades [confirmRebalance] would place. Read-only. */
    fun requestRebalance(id: String) {
        _state.update { it.copy(errorMessage = null) }
        scope.launch {
            try {
                val preview = rebalancePie.preview(id)
                _state.update { it.copy(rebalancePreview = preview) }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                _state.update { it.copy(errorMessage = tr(L10n.Key.CouldntPlaceOrder)) }
            }
        }
    }

    /** Executes the previewed orders, then clears the preview and refreshes rows/detail. */
    fun confirmRebalance(id: String) {
        _state.update { it.copy(errorMessage = null) }
        scope.launch {
            try {
                rebalancePie.execute(id, calendar.tradingDay(nowEpochSeconds()), nowEpochSeconds())
                _state.update { it.copy(rebalancePreview = null) }
                reloadRows()
                if (_state.value.detail?.pieId == id) loadDetail(id)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                _state.update { it.copy(errorMessage = tr(L10n.Key.CouldntPlaceOrder)) }
            }
        }
    }

    fun deletePie(id: String) {
        _state.update { it.copy(rebalancePreview = null) }
        scope.launch {
            deletePieUseCase.execute(id)
            if (_state.value.detail?.pieId == id) {
                _state.update { it.copy(detail = null) }
            }
            reloadRows()
        }
    }

    // MARK: - Row / detail construction

    private suspend fun reloadRows() {
        val pies = loadPies.execute()
        val symbols = pies.flatMap { pie -> pie.slices.map { it.symbol } }.distinct()
        val quotes = fetchQuotesSafely(symbols)
        _state.update { it.copy(rows = pies.map { pie -> buildRow(pie, quotes) }) }
    }

    /** Shared by [openDetail], [contributeNow], and [confirmRebalance]'s "reopen the
     *  currently-open detail if it's this pie" step — a private suspend helper (rather than
     *  those callers re-invoking the public, self-launching [openDetail]) so the reopen runs
     *  sequentially within the SAME coroutine as the mutation that triggered it, instead of a
     *  second, independently-scheduled `scope.launch`. */
    private suspend fun loadDetail(id: String) {
        val pie = loadPies.execute().firstOrNull { it.id == id }
        if (pie == null) {
            _state.update { it.copy(detail = null) }
            return
        }
        val quotes = fetchQuotesSafely(pie.slices.map { it.symbol })
        _state.update { it.copy(detail = buildDetail(pie, quotes)) }
    }

    private fun buildRow(pie: Pie, quotes: Map<String, Quote>): PieRowUi {
        val currentValuesMap = currentValues(pie, quotes)
        val totalValue = currentValuesMap.values.fold(Money(ZERO)) { acc, m -> acc + m }
        val drift = PieMath.drift(currentValuesMap, pie.slices)
        val maxDriftPP = drift.values.maxOfOrNull { it.abs() } ?: ZERO
        return PieRowUi(
            id = pie.id,
            name = pie.name,
            currentValue = totalValue,
            nextContributionLabel = nextContributionLabel(pie),
            maxDriftPP = maxDriftPP,
            sliceWeights = pie.slices.map { it.symbol to it.targetWeightPP },
        )
    }

    private fun buildDetail(pie: Pie, quotes: Map<String, Quote>): PieDetailUi {
        val currentValuesMap = currentValues(pie, quotes)
        // Same reduction `buildRow` uses for this pie's list-card `currentValue` — one
        // source of truth for "how much is this pie worth" shared by the list and detail.
        val totalMoney = currentValuesMap.values.fold(Money(ZERO)) { acc, m -> acc + m }
        val totalValue = totalMoney.amount
        val driftBySymbol = PieMath.drift(currentValuesMap, pie.slices)
        val sliceDetails = pie.slices.map { slice ->
            val value = currentValuesMap[slice.symbol] ?: Money(ZERO)
            val actual = if (totalValue > ZERO) {
                value.amount.divide(totalValue, MONEY_MATH) * ONE_HUNDRED
            } else {
                ZERO
            }
            PieSliceDetailUi(
                symbol = slice.symbol,
                assetKind = slice.assetKind,
                targetWeight = slice.targetWeightPP,
                actualWeight = actual,
                drift = driftBySymbol[slice.symbol] ?: ZERO,
                currentValue = value,
            )
        }
        return PieDetailUi(
            pieId = pie.id,
            name = pie.name,
            slices = sliceDetails,
            activity = pie.activity,
            schedule = pie.schedule,
            totalValue = totalMoney,
        )
    }

    /** Prices `pie`'s ledger against `quotes`; a symbol with no successful quote is simply
     *  omitted (mirrors [PieMath]'s own "0 if missing" treatment). */
    private fun currentValues(pie: Pie, quotes: Map<String, Quote>): Map<String, Money> {
        val result = LinkedHashMap<String, Money>()
        for (slice in pie.slices) {
            val quote = quotes[slice.symbol] ?: continue
            val quantity = pie.quantityOf(slice.symbol)
            result[slice.symbol] = Money(quantity * quote.price.amount, quote.price.currencyCode)
        }
        return result
    }

    /** Batch quote fetch for row/detail pricing. Unlike [PortfolioViewModel]'s
     *  poll-and-keep-last-good `refreshQuotes`, there is no "last good" cache for pie
     *  pricing — on any failure this degrades to an empty map, which [currentValues] already
     *  treats identically to "no quote for this symbol" (mirrors Swift's per-symbol
     *  `Result`-based `fetchQuotes`, which never throws). */
    private suspend fun fetchQuotesSafely(symbols: List<String>): Map<String, Quote> {
        if (symbols.isEmpty()) return emptyMap()
        return try {
            fetchMarketQuotes.execute(symbols).associateBy { it.symbol }
        } catch (e: CancellationException) {
            throw e
        } catch (e: QuoteError) {
            emptyMap()
        }
    }

    private fun nextContributionLabel(pie: Pie): String? {
        val schedule = pie.schedule ?: return null
        val epochDay = PieSchedule.parseDay(schedule.nextDueDay) ?: return null
        val formatted = LocalDate.ofEpochDay(epochDay).format(DUE_DAY_FORMATTER)
        return trf(L10n.Key.NextContributionFormat, formatted)
    }

    private companion object {
        val ONE_HUNDRED: BigDecimal = BigDecimal.parseString("100")
        val DUE_DAY_FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("MMM d", Locale.US)
    }
}
